src/workerd/api/actor.h

```diff
@@ -13,27 +13,95 @@
 #include <capnp/compat/http-over-capnp.h>
 #include <workerd/api/http.h>
 #include <workerd/jsg/jsg.h>
+#include <workerd/jsg/function.h>
 
 namespace workerd {
   template <typename T> class IoOwn;
 }
 
 namespace workerd::api {
 
+enum IsInHouse {
+  YES,
+  NO,
+};
+
+// This class is the client RPC stub for Durable Objects. It intercepts dynamic method calls, then
+// sends the method name and arguments to the remote Durable Object, returning the response it
+// receives.
+//
+// We don't want name interception on Fetcher itself -- at least not yet -- since not all Fetcher
+// implementations need it. Name interception is off by default; to turn it on, set the JsRpcStub
+// compatibility flag.
+class DurableObjectRpcStub : public Fetcher, public jsg::NamedIntercept {
+public:
+  DurableObjectRpcStub(
+      IoOwn<OutgoingFactory> outgoingFactory,
+      RequiresHostAndProtocol requiresHost,
+      IsInHouse inHouse,
+      CompatibilityFlags::Reader flags)
+    : Fetcher(kj::mv(outgoingFactory), requiresHost, (inHouse == YES)), flags(flags) {}
+
+  bool isRegisteredName(kj::StringPtr name) {
+    // We don't need to check `fetch` or `constructor` because `fetch` is known at compile time and
+    // `constructor` is a JS keyword; neither uses name interception. This needs to be updated
+    // whenever a new "registered name", aka handler, is added to Durable Objects.
+    if (name == "alarm"_kjc ||
+        name == "webSocketMessage"_kjc ||
+        name == "webSocketClose"_kjc ||
+        name == "webSocketError"_kjc) {
+      return true;
+    }
+    return false;
+  }
+
+  kj::Maybe<jsg::JsValue> getNamed(jsg::Lock& js, kj::StringPtr name) override {
+    if (flags.getJsRpcStub()) {
+      JSG_REQUIRE(!isRegisteredName(name), Error, "cannot call registered method '", name, "'");
+
+      // Named intercept is enabled, this means we won't default to legacy behavior.
+      // TODO(soon): The server side PR will fill this function with the RPC call and other stuff,
```

Comment:
"Client PR" would be better said as: "Give the DO stub the ability to intercept method names", but then don't actually do anything because all the JS RPC stuff is in the other PR.

The only client side change we make in the other ("server side") PR is filling out the `getNamed()` method to use the Cap'n Proto stub https://github.com/cloudflare/workerd/pull/1028/commits/898c892d4fc31b7f5d1d9d0238951cb5515cc8f1#diff-a52f294aa5682a0d24159ba162f5dbbf0fe56e4d87e7f3e87b4efdb833299e8dR64

Mostly thought this would be easier to review (separating JSG/stub changes from RPC change) and client could land first (less risk of breaking something).